File: daemons/macos/src/Handlers/ChatHandler.swift
import Foundation
import Network

enum ChatHandler {
    static func start(_ request: HTTPRequest, params: [String: String]) -> HTTPResponse {
        if let sessionId = params["id"],
            let body = try? JSONSerialization.jsonObject(with: request.body) as? [String: Any],
            let path = body["path"] as? String,
            let prompt = body["prompt"] as? String
        {
            let images = (body["images"] as? [[String: String]]) ?? []
            let existsOnServer = (body["existsOnServer"] as? Bool) ?? false
            let model = body["model"] as? String
            let effort = body["effort"] as? String
            let permissionMode = body["permissionMode"] as? String
            let provider = body["provider"] as? String
            let threadId = body["threadId"] as? String
            if let reviewTarget = body["reviewTarget"], provider != "codex" || !CodexReviewTarget.valid(reviewTarget) {
                return HTTPResponse.json(
                    400,
                    ["error": "Provide a valid Codex review target: uncommittedChanges, baseBranch, commit, or custom."]
                )
            }
            if let shellCommand = body["shellCommand"],
                provider != "codex" || !CodexShellCommand.valid(shellCommand)
                    || body["reviewTarget"] != nil || !images.isEmpty
                    || !(body["skills"] as? [[String: String]] ?? []).isEmpty
                    || !(body["mentions"] as? [[String: String]] ?? []).isEmpty
            {
                return HTTPResponse.json(
                    400, ["error": "Provide one Codex command without images, references or a review target."])
            }
            if provider == "codex", let id = threadId ?? CodexSessionStore.shared.threadId(for: sessionId),
                CodexClient.shared.isThreadActive(threadId: id)
            {
                return HTTPResponse.json(
                    409, ["error": "This task is running on the host. Wait for it to finish before continuing."])
            }
            if CodexHandler.isMutating(sessionId) || RunnerManager.shared.isRunning(sessionId: sessionId) {
                return HTTPResponse.json(409, ["error": "session_already_running"])
            }
            if let provider, !["codex", "claude"].contains(provider) {
                return HTTPResponse.json(400, ["error": "unknown_provider"])
            }
            #if DEBUG
            NSLog(
                "[ChatHandler] start sessionId=\(sessionId) path=\(path) existsOnServer=\(existsOnServer) model=\(model ?? "nil") effort=\(effort ?? "nil") permissionMode=\(permissionMode ?? "nil") promptChars=\(prompt.count) images=\(images.count)"
            )
            #endif
            return HTTPResponse.stream { connection in
                RunnerManager.shared.start(
                    sessionId: sessionId, path: path, prompt: prompt, images: images,
                    existsOnServer: existsOnServer, model: model, effort: effort,
                    permissionMode: permissionMode, provider: provider, threadId: threadId,
                    skills: body["skills"] as? [[String: String]] ?? [],
                    mentions: body["mentions"] as? [[String: String]] ?? [],
                    projectId: body["projectId"] as? String,
                    reviewTarget: body["reviewTarget"] as? [String: Any],
                    shellCommand: body["shellCommand"] as? String,
                    connection: connection
                )
            }
        }
        return HTTPResponse.json(400, ["error": "bad_request"])
    }

    static func resume(_ request: HTTPRequest, params: [String: String]) -> HTTPResponse {
        if let sessionId = params["id"] {
            let afterSeq = Int(request.query["after_seq"] ?? "") ?? -1
            #if DEBUG
            NSLog("[ChatHandler] resume sessionId=\(sessionId) afterSeq=\(afterSeq)")
            #endif
            return HTTPResponse.stream { connection in
                let attached = RunnerManager.shared.resumeIfExists(
                    sessionId: sessionId, afterSeq: afterSeq, connection: connection
                )
                if !attached,
                    !CodexJournalReplay.resume(sessionId: sessionId, afterSeq: afterSeq, connection: connection),
                    !SessionJSONLReplay.replay(sessionId: sessionId, to: connection)
                {
                    connection.cancel()
                }
            }
        }
        return HTTPResponse.json(400, ["error": "bad_request"])
    }

    static func abort(_ request: HTTPRequest, params: [String: String]) -> HTTPResponse {
        if let sessionId = params["id"] {
            #if DEBUG
            NSLog("[ChatHandler] abort sessionId=\(sessionId)")
            #endif
            if RunnerManager.shared.abort(sessionId: sessionId) {
                return HTTPResponse.json(200, ["ok": true, "aborted": true])
            }
            return CodexControlHandler.abort(sessionId: sessionId)
        }
        return HTTPResponse.json(400, ["error": "bad_request"])
    }

    static func steer(_ request: HTTPRequest, params: [String: String]) -> HTTPResponse {
        guard request.body.count <= 262_144, request.query.isEmpty,
            let sessionId = params["id"],
            let body = try? JSONSerialization.jsonObject(with: request.body) as? [String: Any],
            body.keys.allSatisfy({ ["prompt", "requestId"].contains($0) }),
            let prompt = body["prompt"] as? String, !prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
            prompt.utf16.count <= 32_768,
            body["requestId"] == nil || (body["requestId"] as? String).flatMap(UUID.init(uuidString:)) != nil
        else { return HTTPResponse.json(400, ["error": "invalid_steering_request"]) }
        let semaphore = DispatchSemaphore(value: 0)
        var result: Result<[String: Any], Error>?
        if RunnerManager.shared.steer(
            sessionId: sessionId, prompt: prompt, requestId: body["requestId"] as? String,
            completion: { response in
                result = response
                semaphore.signal()
            })
        {
            if semaphore.wait(timeout: .now() + 21) == .success, let result {
                switch result {
                case .success: return HTTPResponse.json(200, ["ok": true])
                case .failure(let error):
                    return HTTPResponse.json(
                        [402, 409, 503].contains((error as NSError).code) ? (error as NSError).code : 502,
                        ["error": error.localizedDescription])
                }
            }
            return HTTPResponse.json(504, ["error": "Codex did not confirm the steering message."])
        }
        return HTTPResponse.json(409, ["error": "no_active_codex_turn"])
    }

    static func respond(_ request: HTTPRequest, params: [String: String]) -> HTTPResponse {
        if let sessionId = params["id"],
            let body = try? JSONSerialization.jsonObject(with: request.body) as? [String: Any],
            let requestId = body["requestId"] as? String,
            let result = body["result"] as? [String: Any]
        {
            let responded = RunnerManager.shared.respond(sessionId: sessionId, requestId: requestId, result: result)
            return HTTPResponse.json(responded ? 200 : 404, ["ok": responded])
        }
        return HTTPResponse.json(400, ["error": "bad_request"])
    }

    static func requests(_ request: HTTPRequest, params: [String: String]) -> HTTPResponse {
        if let sessionId = params["id"] {
            return HTTPResponse.json(
                200,
                [
                    "requests": RunnerManager.shared.requests(sessionId: sessionId),
                    "agentAttention": RunnerManager.shared.agentAttention(sessionId: sessionId),
                ])
        }
        return HTTPResponse.json(400, ["error": "bad_request"])
    }
}

File: daemons/linux/src/Handlers/ChatHandler.js
import { validReviewTarget } from '../Codex/CodexReviewTarget.js'
import { codexCompaction } from '../Codex/CodexCompaction.js'
import HTTPResponse from '../Networking/HTTPResponse.js'
import path from 'node:path'
import os from 'node:os'
import { runnerManager } from '../RunnerManager.js'
import { replay } from './SessionJSONLReplay.js'
import { codexSessions } from '../Codex/CodexSessions.js'
import { codexClient } from '../Codex/CodexClient.js'

function parsedBody(request) {
  try {
    return JSON.parse(request.body.toString('utf8'))
  } catch {
    return null
  }
}

export function start(request, params) {
  const body = parsedBody(request)
  if (params.id && codexSessions.busy(params.id)) { return HTTPResponse.json(409, { error: 'session_update_pending', message: 'A task update is in progress. Retry when it finishes.' }) }
  if (body?.shellCommand !== undefined && (body.provider !== 'codex' || typeof body.shellCommand !== 'string' || !body.shellCommand.trim() || body.shellCommand.length > 16384 || body.reviewTarget !== undefined || ['images', 'skills', 'mentions'].some((key) => body[key] !== undefined && (!Array.isArray(body[key]) || body[key].length > 0)))) {
    return HTTPResponse.json(400, { error: 'Provide a Codex shell command up to 16384 characters without images, references, or a review target.' })
  }
  if (body?.reviewTarget !== undefined && (body.provider !== 'codex' || !validReviewTarget(body.reviewTarget))) {
    return HTTPResponse.json(400, { error: 'Provide a valid Codex review target: uncommittedChanges, baseBranch, commit, or custom.' })
  }
  if (body?.provider && !['claude', 'codex'].includes(body.provider)) {
    return HTTPResponse.json(400, { error: 'unsupported_provider' })
  }
  if (body?.provider === 'codex' && (codexClient.activeTurns.has(body.threadId || codexSessions.read(params.id)?.threadId) || codexCompaction.busy(body.threadId || codexSessions.read(params.id)?.threadId) || runnerManager.runners.has(params.id.toLowerCase()) || [...runnerManager.runners.values()].some((runner) => runner.threadId && runner.threadId === (body.threadId || codexSessions.read(params.id)?.threadId)))) {
    return HTTPResponse.json(409, { error: 'turn_already_running', message: 'Steer or interrupt the active turn before starting another.' })
  }
  if (params.id && typeof body?.path === 'string' && body.path.length > 0 && typeof body?.prompt === 'string') {
    return HTTPResponse.stream(200, 'application/x-ndjson', {}, (response) => {
      runnerManager.start({
        sessionId: params.id,
        provider: body.provider,
        reviewTarget: body.reviewTarget,
        shellCommand: body.shellCommand,
        skills: Array.isArray(body.skills) ? body.skills : [],
        mentions: Array.isArray(body.mentions) ? body.mentions : [],
        projectId: typeof body.projectId === 'string' ? body.projectId : null,
        threadId: typeof body.threadId === 'string' ? body.threadId : null,
        path: body.path === '~' ? os.homedir() : body.path.startsWith('~/') ? path.join(os.homedir(), body.path.slice(2)) : path.resolve(body.path),
        prompt: body.prompt,
        images: Array.isArray(body.images) ? body.images : [],
        existsOnServer: body.existsOnServer === true,
        model: typeof body.model === 'string' ? body.model : null,
        effort: typeof body.effort === 'string' ? body.effort : null,
        permissionMode: typeof body.permissionMode === 'string' ? body.permissionMode : null,
        response
      })
    })
  }
  return HTTPResponse.json(400, { error: 'bad_request' })
}

export function resume(request, params) {
  if (params.id) {
    return HTTPResponse.stream(200, 'application/x-ndjson', {}, (response) => {
      const parsed = Number.parseInt(request.query.after_seq || '-1', 10)
      const afterSeq = Number.isNaN(parsed) ? -1 : parsed
      const attached = runnerManager.resumeIfExists(params.id, afterSeq, response)
      if (!attached && !codexSessions.replay(params.id, response) && !replay(params.id, response)) {
        response.end()
      }
    })
  }
  return HTTPResponse.json(400, { error: 'bad_request' })
}

export async function abort(request, params) {
  if (params.id) {
    if (runnerManager.abort(params.id)) { return HTTPResponse.json(200, { ok: true, aborted: true }) }
    const saved = codexSessions.read(params.id)
    if (!saved?.threadId || saved.provider !== 'codex') { return HTTPResponse.json(200, { ok: true, aborted: false }) }
    return codexClient.request('thread/read', { threadId: saved.threadId, includeTurns: true }).then(async (result) => {
      if (result.thread?.id !== saved.threadId) { return HTTPResponse.json(409, { error: 'The host returned a different task. Refresh before stopping it.' }) }
      if (result.thread.status?.type !== 'active') { return HTTPResponse.json(200, { ok: true, aborted: false }) }
      const turn = result.thread.turns?.at(-1)
      if (turn?.status !== 'inProgress' || typeof turn.id !== 'string' || !turn.id) { return HTTPResponse.json(409, { error: 'The active turn is not available yet. Refresh the task and try again.' }) }
      await codexClient.request('turn/interrupt', { threadId: saved.threadId, turnId: turn.id })
      return HTTPResponse.json(200, { ok: true, aborted: true, threadId: saved.threadId, turnId: turn.id })
    }).catch(() => HTTPResponse.json(502, { error: 'The host could not interrupt this task. Refresh its status and try again.' }))
  }
  return HTTPResponse.json(400, { error: 'bad_request' })
}
